### PR TITLE
Various cleanups in prometheus-metrics-core

### PR DESCRIPTION
**Pull request by mimaison**
Also added unit tests for CounterWithCallback, GaugeWithCallback, Summary and SummaryWithCallback.

### CODE REVIEW

File: prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/Histogram.java [hunk 1]
```diff
@@ -481,8 +481,8 @@ private boolean maybeReset() {
       nativeZeroCount.reset();
       count.reset();
       sum.reset();
-      for (int i = 0; i < classicBuckets.length; i++) {
-        classicBuckets[i].reset();
+      for (LongAdder classicBucket : classicBuckets) {
```
Comment (zeitlinger) [inline prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/Histogram.java:484]: the performance is not the same - but I guess it's ok here

File: prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/CounterWithCallbackTest.java [hunk 2]
```diff
@@ -1,6 +1,42 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Test;
+
 public class CounterWithCallbackTest {
 
-  // TODO :). Anyway, callbacks are implicitly covered by the JVM metrics tests as well.
+  @Test
+  public void testCounter() {
```
Comment (zeitlinger) [inline prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/CounterWithCallbackTest.java:14]: wdyt of using assertj for new tests?

File: prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/Histogram.java (hunk 1 above)
Comment (mimaison) [reply prometheus-metrics-core/src/main/java/io/prometheus/metrics/core/metrics/Histogram.java:484]: My understanding is that on a modern JVM the performance difference should be negligible.

File: prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/CounterWithCallbackTest.java (hunk 2 above)
Comment (mimaison) [reply prometheus-metrics-core/src/test/java/io/prometheus/metrics/core/metrics/CounterWithCallbackTest.java:14]: As I'm not a maintainer on this project, I wasn't sure about adding new dependencies. I see you're doing the switch to assertj in https://github.com/prometheus/client_java/pull/1092

**Pull request merged**